pkg/osquery/tables/gsettings/gsettings.go

```diff
@@ -0,0 +1,282 @@
+package gsettings
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	osquery "github.com/kolide/osquery-go"
+	"github.com/kolide/osquery-go/plugin/table"
+	"github.com/pkg/errors"
+)
+
+type Table struct {
+	client *osquery.ExtensionManagerClient
+	logger log.Logger
+}
+
+func TablePlugin(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+	columns := []table.ColumnDefinition{
+		// TODO: maybe need to add 'path' for relocatable schemas..
+		table.TextColumn("domain"),
+		table.TextColumn("key"),
+		table.TextColumn("type"),
+		table.TextColumn("value"),
+		table.TextColumn("description"),
+	}
+
+	t := &Table{
+		client: client,
+		logger: logger,
+	}
+
+	return table.NewPlugin("kolide_gsettings", columns, t.generate)
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+
+	domainQ, ok := queryContext.Constraints["domain"]
+	if !ok || len(domainQ.Constraints) == 0 {
+		return nil, errors.New("the kolide_gsettings table requires a domain to be specified")
+	}
+
+	for _, domainConstraint := range domainQ.Constraints {
+		// TODO: this doesn't handle wildcards or partial matches..
+		osqueryResults, err := t.gsettings(ctx, domainConstraint.Expression)
+		if err != nil {
+			continue
+		}
+
+		for _, row := range osqueryResults {
+			results = append(results, row)
+		}
+	}
+	return results, nil
+}
+
+func (t *Table) gsettings(ctx context.Context, domain string) ([]map[string]string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, "/usr/bin/gsettings", "list-recursively", domain)
+	dir, err := ioutil.TempDir("", "osq-gsettings")
+	if err != nil {
+		return nil, errors.Wrap(err, "mktemp")
+	}
+	defer os.RemoveAll(dir)
+
+	if err := os.Chmod(dir, 0755); err != nil {
+		return nil, errors.Wrap(err, "chmod")
+	}
+	cmd.Dir = dir
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+	cmd.Stdout, cmd.Stderr = stdout, stderr
+
+	if err := cmd.Run(); err != nil {
+		level.Info(t.logger).Log(
+			"msg", "Error running gsettings",
+			"stderr", strings.TrimSpace(stderr.String()),
+			"stdout", strings.TrimSpace(stdout.String()),
+			"err", err,
+		)
+		return nil, errors.Wrap(err, "running osquery")
+	}
+	osqueryResults := t.parse(stdout)
+	for _, row := range osqueryResults {
+		d, err := t.gsettingsDescribe(ctx, row["key"], row["domain"])
+		if err != nil {
+			continue
+		}
+		row["description"] = d.Description
+		row["type"] = d.Type
+	}
+
+	return osqueryResults, nil
+}
+
+type datatype struct {
+	raw string
+}
+
+type keyDescription struct {
+	Description string
+	Type        string
+}
+
+func (t *Table) gsettingsDescribe(ctx context.Context, key, domain string) (keyDescription, error) {
```

Comment:
I think I don't like the N+1. 

I think there are 2 different functions. One is to use `list-recursively` when selecting a large block. The other is to get info about a specific setting.

I could believe doing that through either a single table that behaves different depending on what you call it with, or doing it though two tables.

Reply:
Yea I thought of adding a column that's only populated if you explicitly select from it, though it feels odd to break `select *`.  

Or (I think this is what you are referring to) we could only list the description column when it's explicitly mentiond, _and_ when there are constraints on the `key` column..

I'm not entirely against a 2nd table, but I think I prefer a single table. Still need to ponder this more

Reply:
I see several models

1. As is. This selects everything. Too many execs
2. two tables. Maybe one to use `list-recursively` to get values. and a `gsettings_descriptions` with a required `key`
3. single table, only return descriptions if the column is selected (this is akin to a hidden column in osquery)
4. single table, only return descriptions if a key is specified. (eg: `select *` gets values, `select * where key in (a, b, c)` would also populate the description

At least this evening, I'm favoring (2) or (4)

Reply:
I kind of like (2) the more I think about it. `gsettings` and `gsettings_metadata` or something, we could have explicit 'type' and 'description' data there.